agent/artifact_downloader.go

```diff
@@ -76,81 +78,112 @@ func (a *ArtifactDownloader) Download() error {
+	}
+
+	for _, artifact := range artifacts {
+		// Create new instance of the artifact for the goroutine
+		// See: http://golang.org/doc/effective_go.html#channels
+		artifact := artifact
+
+		p.Spawn(func() {
+			// Convert windows paths to slashes, otherwise we get a literal
+			// download of "dir/dir/file" vs sub-directories on non-windows agents
+			path := artifact.Path
+			if runtime.GOOS != `windows` {
+				path = strings.Replace(path, `\`, `/`, -1)
+			}
+
+			// Handle downloading from S3, GS, or RT
+			var err error
+			switch {
+			case strings.HasPrefix(artifact.UploadDestination, "s3://"):
+				bucketName, _ := ParseS3Destination(artifact.UploadDestination)
+				err = NewS3Downloader(a.logger, S3DownloaderConfig{
+					S3Client:    s3Clients[bucketName],
+					Path:        path,
+					S3Path:      artifact.UploadDestination,
+					Destination: downloadDestination,
+					Retries:     5,
+					DebugHTTP:   a.conf.DebugHTTP,
+				}).Start()
+			case strings.HasPrefix(artifact.UploadDestination, "gs://"):
+				err = NewGSDownloader(a.logger, GSDownloaderConfig{
+					Path:        path,
+					Bucket:      artifact.UploadDestination,
+					Destination: downloadDestination,
+					Retries:     5,
+					DebugHTTP:   a.conf.DebugHTTP,
+				}).Start()
+			case strings.HasPrefix(artifact.UploadDestination, "rt://"):
+				err = NewArtifactoryDownloader(a.logger, ArtifactoryDownloaderConfig{
+					Path:        path,
+					Repository:  artifact.UploadDestination,
+					Destination: downloadDestination,
+					Retries:     5,
+					DebugHTTP:   a.conf.DebugHTTP,
+				}).Start()
+			default:
+				err = NewDownload(a.logger, http.DefaultClient, DownloadConfig{
+					URL:         artifact.URL,
+					Path:        path,
+					Destination: downloadDestination,
+					Retries:     5,
+					DebugHTTP:   a.conf.DebugHTTP,
+				}).Start()
+			}
+
+			// If the downloaded encountered an error, lock
+			// the pool, collect it, then unlock the pool
+			// again.
+			if err != nil {
+				a.logger.Error("Failed to download artifact: %s", err)
+
+				p.Lock()
+				errors = append(errors, err)
+				p.Unlock()
+			}
+		})
+	}
+
+	p.Wait()
+
+	if len(errors) > 0 {
+		return fmt.Errorf("There were errors with downloading some of the artifacts")
+	}
+
+	return nil
+}
+
+// We want to have as few S3 clients as possible, as creating them is kind of an expensive operation
+// But it's also theoretically possible that we'll have multiple artifacts with different S3 buckets, and each
+// S3Client only applies to one bucket, so we need to store the S3 clients in a map, one for each bucket
+func (a *ArtifactDownloader) generateS3Clients(artifacts []*api.Artifact) (map[string]*s3.S3, error) {
+	s3Clients := map[string]*s3.S3{}
+
+	for _, artifact := range artifacts {
+		if !strings.HasPrefix(artifact.UploadDestination, "s3://") {
+			continue
 		}
 
-		p.Wait()
+		bucketName, _ := ParseS3Destination(artifact.UploadDestination)
+		if _, has := s3Clients[bucketName]; !has {
```

Comment:
This go-ism took me way too long to understand. But I think I got there. This runs the block if `bucketName` isn't already in the map.

Reply:
yeah, something like `maps.HasKey(s3Clients, bucketName)` would be too simple and readable aye

(`maps.HasKey` does not exist)

Reply:
I tend to use `ok` e.g. `if _, ok := s3Clients[bucketName]; !ok {` which is no less mysterious but I thought it was more Go-idiomatic, but now I'm not sure.

Reply:
[This](https://go.dev/doc/effective_go#maps) implies `ok` is idiomatic:

![image](https://user-images.githubusercontent.com/15759/182295955-6ef2bd13-3732-489b-bde5-58bb008a23fa.png)

----

But the next paragraph implies it can be any suitable word 🤷‍♂️
![image](https://user-images.githubusercontent.com/15759/182295989-8181bedb-abb1-4b5e-a877-382b25433f55.png)

Reply:
yeah, i've mostly used `ok` in the past, but `has` is fairly common in the agent codebase:
```
cache-s3-creds (~/b/agent) rg "has :="
tracetools/propagate.go
37:	s, has := env[EnvVarTraceContextKey]

clicommand/agent_start.go
698:		if _, has := tracetools.ValidTracingBackends[cfg.TracingBackend]; !has {

agent/pipeline_parser.go
65:	if tracing, has := p.Env.Get(tracetools.EnvVarTraceContextKey); has {
67:		if envMap, has := mapSliceItem("env", pipeline); has {

agent/artifact_downloader.go
178:		if _, has := s3Clients[bucketName]; !has {

bootstrap/tracing.go
44:		if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
193:	if attemptStr, has := env.Get("BUILDKITE_RETRY_COUNT"); has {
200:	if parallelStr, has := env.Get("BUILDKITE_PARALLEL_JOB"); has {
206:	rebuiltFromID, has := env.Get("BUILDKITE_REBUILT_FROM_BUILD_NUMBER")
211:	triggeredFromID, has := env.Get("BUILDKITE_TRIGGERED_FROM_BUILD_ID")
216:	jobLabel, has := env.Get("BUILDKITE_LABEL")
221:	jobKey, has := env.Get("BUILDKITE_STEP_KEY")
```

and i kinda like it from a readability standpoint - i think the `val, ok := someVar.(string)` convention makes a bunch of sense because the `ok` is "did the type cast work ok", but in the case of `val, ok := someMap[key]`, it's not really whether something is ok or not, it's whether it's present.